tessel/src/lib.rs

```diff
@@ -82,86 +78,146 @@ pub struct PortGroup {
 /// use tessel::Port;
 /// ```
 pub struct Port {
-    // Path of the domain socket.
     socket: Arc<Mutex<PortSocket>>,
-    pins: HashMap<usize, Mutex<()>>,
 }
 
-pub struct Pin<'a> {
-    index: usize,
-    _guard: MutexGuard<'a, ()>,
-    socket: Arc<Mutex<PortSocket>>,
-}
+impl Port {
+    pub fn new(path: &str) -> Port {
+        // Create and return the port struct
+        Port {
+            socket: Arc::new(Mutex::new(PortSocket::new(path))),
+        }
+    }
 
-impl<'a> Pin<'a> {
-    pub fn output(&mut self, value: bool) -> io::Result<()> {
-        let mut sock = self.socket.lock().unwrap();
-        if value {
-            sock.write_command(Command::GpioHigh(self.index as u8))
-        } else {
-            sock.write_command(Command::GpioLow(self.index as u8))
+    pub fn pins(&mut self) -> (Pin, Pin, Pin) {
+        (
+            Pin::new(5, self.socket.clone()),
+            Pin::new(6, self.socket.clone()),
+            Pin::new(7, self.socket.clone()),
+        )
+    }
+
+    pub fn i2c<'b>(self) -> (I2cPort<'b>, Gpio<'b>) {
+        let mut available = BTreeSet::new();
+        for i in 2..8 {
+            available.insert(i);
         }
+        (I2cPort::new(self.socket.clone()), Gpio::new(self.socket.clone(), available))
     }
 }
 
-impl Port {
-    pub fn new(path: &str) -> Port {
-        let mut pins = HashMap::new();
-        for i in 0..8 {
-            pins.insert(i, Mutex::new(()));
-        }
+/// Gpio is a selection of pins.
+#[allow(dead_code)]
+pub struct Gpio<'a> {
+    socket: Arc<Mutex<PortSocket>>,
+    available: BTreeSet<usize>,
+    _phantom: PhantomData<&'a Port>,
+}
 
+impl<'a> Gpio<'a> {
+    pub fn new<'b>(socket: Arc<Mutex<PortSocket>>, available: BTreeSet<usize>) -> Gpio<'b> {
```

Comment:
Rust newbie question: Why does the `new` function return a tuple with a different lifetime than the struct?